src/script/interpreter.cpp

```diff
@@ -1946,6 +2074,180 @@ bool GenericTransactionSignatureChecker<T>::CheckDefaultCheckTemplateVerifyHash(
+
+    // If this is an unauthenticated recovery, ensure that the only other
+    // output is an ephemeral anchor (by policy).
+    if (flags & SCRIPT_VERIFY_VAULT_UNAUTH_RECOVERY_STRUCTURE && !is_authed_recovery) {
+        const size_t optional_ea_idx{recovery_vout_idx == 0 ? 1U : 0U};
+        if (num_vout != 2 || !IsOutputEphemeralAnchor(vout[optional_ea_idx])) {
+            return SCRIPT_ERR_VAULT_BAD_RECOVERY_OUTPUTS;
+        }
+    }
+
+    execdata.AddDeferredVaultRecoveryCheck(recovery_vout_idx, this->amount);
+
+    return std::nullopt;
+}
+
+template <class T>
+std::optional<ScriptError> GenericTransactionSignatureChecker<T>::CheckVaultTrigger(
+    ScriptExecutionData& execdata,
+    const size_t trigger_out_idx,
+    const std::optional<size_t> maybe_revault_out_idx,
+    CScript flu_script,
+    unsigned int flags,
+    ScriptError* serror) const
+{
+    assert(execdata.m_curr_witversion);
+
+    if (!this->txdata) {
+        HandleMissingData(m_mdb);
+        return SCRIPT_ERR_UNKNOWN_ERROR;
+    }
+    const auto& txd{*this->txdata};
+    const auto& vout{this->txTo->vout};
+
+    assert(txd.m_spent_outputs_ready);
+    const CScript& vault_spk = txd.m_spent_outputs[this->nIn].scriptPubKey;
+
+    if (trigger_out_idx >= vout.size()) {
+        return SCRIPT_ERR_UNVAULT_MISMATCH;
+    }
+
+    // Ensure that the output creating the unvault trigger matches the
+    // expected scriptPubKey.
+    const auto& value_out = vout[trigger_out_idx];
+
+    int trigger_witversion;
+    valtype trigger_witprogram;
+    const bool is_unvault_output_wit =
+        value_out.scriptPubKey.IsWitnessProgram(trigger_witversion, trigger_witprogram);
+
+    if (!is_unvault_output_wit || trigger_witversion < 1 ||
+            trigger_witversion != *execdata.m_curr_witversion) {
+        return SCRIPT_ERR_UNVAULT_INCOMPAT_OUTPUT_TYPE;
+    }
+    else if (trigger_witversion == 1) {
+        assert(execdata.m_internal_key);
+        assert(execdata.m_taproot_control.size() > 0);
+
+        const valtype& control = execdata.m_taproot_control;
+        const valtype val_sPK{
+            value_out.scriptPubKey.begin(), value_out.scriptPubKey.end()};
+
+        if (val_sPK.size() != 34) {
+            return SCRIPT_ERR_UNVAULT_MISMATCH;
+        }
+        const XOnlyPubKey val_pk{Span{val_sPK}.subspan(2)};
+
+        auto tapleaf_hash = ComputeTapleafHash(
+            control[0] & TAPROOT_LEAF_MASK, flu_script);
+        const uint256 merkle_root = ComputeTaprootMerkleRoot(control, tapleaf_hash);
+        const auto expected_pk = (*execdata.m_internal_key).CreateTapTweak(&merkle_root);
+
+        // Ensure the currently executing OP_VAULT tapleaf is substituted for the
+        // expected replacement.
+        if (!(expected_pk && val_pk == expected_pk->first)) {
+            return SCRIPT_ERR_UNVAULT_MISMATCH;
+        }
+    }
+    // Append the handling of future witness versions with `else if`s here.
+    else {
```

Comment:
this doesn't match the spec(and is unsafe, otherwise people can just steal from a vault to a future output)

"If the scriptPubKey of triggerOut is not a witness program of the same version and same tapleaf version as the currently executing script, script execution MUST fail and terminate immediately."

Reply:
ok, in that case I think this is dead code, and something like https://github.com/bitcoin-inquisition/bitcoin/pull/21/files#r1224396136 would help make this clear/easier to test